syntax = "proto3";

package chat.v1;

import "google/protobuf/timestamp.proto";
import "google/protobuf/empty.proto";

service ChatService {
    rpc Send(SendRequest) returns (google.protobuf.Empty) {}
    rpc React(ReactionRequest) returns (google.protobuf.Empty) {}
    rpc Fetch(google.protobuf.Empty) returns (stream FetchResponse) {}
    rpc Subscribe(stream SubscribeRequest) returns (stream SubscribeResponse) {}
    rpc HealthCheck(google.protobuf.Empty) returns (HealthCheckResponse) {}
}

message SendRequest {
    string content = 1;
}

message ReactionRequest {
    string message_id = 1;
}

message Reaction {
    string user_name = 1;
}

message Message {
    string message_id = 1;
    string user_name = 2;
    string content = 3;
    repeated Reaction reactions = 4;
    string created_time = 5;
}

message FetchResponse {
    Message msg = 1;
}

message SubscribeRequest {
    optional string event_id = 1;
}

message SubscribeResponse {
    optional string event_id = 1;
    optional string type = 2;
    optional string user_id = 3;
    optional string object_id = 4;
}

message HealthCheckResponse {
    string status = 1;
}
